include/odp/api/spec/packet.h

```diff
@@ -1979,6 +1989,147 @@ int8_t odp_packet_shaper_len_adjust(odp_packet_t pkt);
 
+/*
+ *
+ * Packet vector handling routines
+ * ********************************************************
+ *
+ */
+
+/**
+ * Get packet vector handle from event
+ *
+ * Converts an ODP_EVENT_PACKET_VECTOR type event to a packet vector handle
+ *
+ * @param ev Event handle
+ * @return Packet vector handle
+ *
+ * @see odp_event_type()
+ */
+odp_packet_vect_t odp_packet_vect_from_event(odp_event_t ev);
+
+/**
+ * Convert packet vector handle to event
+ *
+ * @param vpkt Packet vector handle
+ *
+ * @return Event handle
+ */
+odp_event_t odp_packet_vect_to_event(odp_packet_vect_t vpkt);
+
+/**
+ * Allocate a packet vector from a packet vector pool
+ *
+ * Allocates a packet vector from the specified packet vector pool.
+ * The pool must have been created with the ODP_POOL_PACKET_VECTOR type.
+ *
+ * @param pool Packet vector pool handle
+ *
+ * @return Handle of allocated packet vector
+ * @retval ODP_PACKET_VECT_INVALID  Packet vector could not be allocated
+ */
+odp_packet_vect_t odp_packet_vect_alloc(odp_pool_t pool);
+
+/**
+ * Free packet vector
+ *
+ * Frees the packet vector into the packet vector pool it was allocated from.
+ *
+ * @param vpkt Packet vector handle
+ */
+void odp_packet_vect_free(odp_packet_vect_t vpkt);
+
+/**
+ * Get packet vector table
+ *
+ * Packet vector table is a array of packets(odp_packet_t) stored
+ * in contiguous memory location. Upon completion of this API,
+ * the implementation returns the packet table pointer in pkts_tbl.
+ *
+ * @param vpkt Packet vector handle
+ * @param [out] pkts_tbl Points to packet vector table
+ *
+ * @return Number of packets available in the vector.
+ *
+ * @note When pktin subsystem is producing the packet vectors,
+ * odp_pktin_vector_prod_config_t::packet_vect_pool shall be used to configure
+ * the pool to form the vector table.
+ * @note The maximum number of packets can hold this vector is defined
+ * by odp_pool_param_t:pktv:max_vect_size. The return value of this
+ * function will not be greater than odp_pool_param_t:pktv:max_vect_size
+ */
+uint32_t odp_packet_vect_tbl(odp_packet_vect_t vpkt, odp_packet_t **pkts_tbl);
+
+/**
+ * Get number of packets in a vector
+ *
+ * @param vpkt Packet vector handle
+ *
+ * @return The number of packets available in the vector
+ */
+uint32_t odp_packet_vect_size(odp_packet_vect_t vpkt);
+
+/**
+ * Set the number of packets stored in a vector
+ *
+ * Update the number of packets stored in a vector. When the application is
+ * producing the packet vector, this function shall be used by the application
+ * to set the number of packets available in this vector.
+ *
+ * @param vpkt Packet vector handle
+ * @param size Number of packets in this vector
+ *
+ * @note The maximum number of packets can hold this vector is defined
+ * by odp_pool_param_t::pktv::max_vect_size. The size value should not be
+ * greater than odp_pool_param_t::pktv::max_vect_size
+ */
+void odp_packet_vect_size_set(odp_packet_vect_t vpkt, uint32_t size);
```

Comment:
if packet_invalids are allowed then this size includes also invalids (?)

Reply:
For the optimized odp_packet_vect_t consumer implementation it is better to have only valid packet handles like other vector processing frameworks.

Reply:
OK. Remember to document that 0 ... size - 1 handles in the vector table needs to be valid handles (when application edits the table).

Reply:
I think we should allow application to write anything in the vector as long as the handles are valid when the vector is passed to ODP (through the size_set function or otherwise). It is just direct memory access after all.

Allowing invalid handles in the vector in size_set() and when converting the vector to an event and passing the event through a queue might be useful in some cases as we discussed, but it will be possible
to add it later without breaking backward compatibility.

Reply:
Yes. It is possible to add it later when we understand the implications without ABI change. For now, I will document it is allowed to have valid handles. The current use case may just to give the table to pktout_send() function to transmit